doc/sphinx/proof-engine/ltac.rst

```diff
@@ -193,74 +134,199 @@ evaluation yields either a term, an integer or a tactic. Intermediate
 results can be terms or integers but the final result must be a tactic
 which is then applied to the focused goals.
 
-There is a special case for ``match goal`` expressions of which the clauses
-evaluate to tactics. Such expressions can only be used as end result of
-a tactic expression (never as argument of a non-recursive local
+There is a special case for :tacn:`match goal` expressions of which the clauses
+evaluate to tactics. Such expressions can only be used as the end result of
+a tactic expression (never as an argument of a non-recursive local
 definition or of an application).
 
-The rest of this section explains the semantics of every construction of
+The rest of this section explains the semantics of each construct of
 |Ltac|.
 
-Sequence
-~~~~~~~~
+.. todo: there are 36 subsections under "Semantics", which seems like far too many
+
+
+Tactics in terms
+~~~~~~~~~~~~~~~~
+
+.. insertprodn term_ltac term_ltac
+
+.. prodn::
+   term_ltac ::= ltac : ( @ltac_expr )
+
+.. todo note that both term_ltac and tactic_arg have this same production.
+
+Substitution
+~~~~~~~~~~~~
+
+.. todo next paragraph: we need a better discussion of substitution.
+   Looks like that also applies to binder_tactics in some form.
+   See https://github.com/coq/coq/pull/12103#discussion_r422105218
+
+:token:`qualid`\s within |Ltac| expressions are used to represent both terms and
+|Ltac| variables.  If the :token:`qualid` corresponds to
+an |Ltac| variable or tactic name, |Ltac| substitutes the value before applying
+the expression.  Generally it's best to choose distinctive names for |Ltac| variables
+that won't clash with term names.  You can use :n:`ltac:(@qualid)` or :n:`(@qualid)`
+to control whether a :token:`qualid` is interpreted as, respectively, an |Ltac|
+variable or a term.
+
+Value types
+~~~~~~~~~~~
+
+Most commonly, the value of an |Ltac| expression is a tactic that can be executed.  While some
+constructs generate integers or strings, there are no operations for combining
+these, for example there's no function to add two integers.  The following constructs allow creation of
+values with types corresponding to various syntactic classes.  Values of all types can be assigned
+to toplevel symbols with the :cmd:`Ltac` command or to local symbols with the :tacn:`let` tactic.
```

Comment:
I think the contradiction is that in the distant past, "a term, an integer, and a tactic" were all the types there were.  Now there are many more types, so that semantics section should be updated.